src/main/java/com/mbcoder/iot/gpslogger/GPS_Logger.java

```diff
@@ -0,0 +1,329 @@
+
+    private Feature latestPosition;  // the latest position read from the GPS
+    private boolean featureUpdated = false;  // flag set to true every time we update the latestPosition from the GPS.
+
+    private Timer loggingTimer; // timer for logging to local geodatabase
+
+    public static void main(String[] args) {
+
+        Application.launch(args);
+    }
+
+    @Override
+    public void start(Stage stage) {
+
+        // set the title and size of the stage and show it
+        stage.setTitle("GPS Logger");
+        stage.setWidth(800);
+        stage.setHeight(100);
+        stage.show();
+
+        // create a JavaFX scene with a stack pane as the root node and add it to the scene
+        StackPane stackPane = new StackPane();
+        Scene scene = new Scene(stackPane);
+        stage.setScene(scene);
+
+        HBox hBox = new HBox();
+        stackPane.getChildren().add(hBox);
+
+        // Button to get offline geodatabase
+        Button btnGetGDB = new Button("Get offline gdb");
+        btnGetGDB.setOnAction(event -> downloadOfflineDB());
+        hBox.getChildren().add(btnGetGDB);
+
+        // Button to open offline geodatabase
+        Button btnOpenGDB = new Button("open offline gdb");
+        btnOpenGDB.setOnAction(event -> {
+            System.out.println("opening gdb");
+            openGeodatabase();
+        });
+        hBox.getChildren().add(btnOpenGDB);
+
+        // Button to sync gps positions collected into the hosted feature service
+        Button btnSync = new Button("Sync data");
+        btnSync.setOnAction(event -> {
+            System.out.println("sync data");
+            syncGeodatabase();
+
+        });
+        hBox.getChildren().add(btnSync);
+
+        // Button to open the serial port and start processing GPS data
+        Button btnStartGPSUpdate = new Button("start GPS");
+        btnStartGPSUpdate.setOnAction(event -> startGPSUpdate());
+        hBox.getChildren().add(btnStartGPSUpdate);
+
+        // Button to start recording the GPS position to offline gdb every 10 seconds
+        Button btnStartLogger = new Button("start logging");
+        btnStartLogger.setOnAction(event -> startLogging());
+        hBox.getChildren().add(btnStartLogger);
+    }
+
+    /**
+     * Method to start logging GPS data into the geodatabase.
+     * <p>
+     * GPS data is reported very frequently, but this method only records new
+     * data at least every 10 seconds to limit the amount of data recorded.
+     */
+    private void startLogging() {
+        loggingTimer = new Timer();
+
+        loggingTimer.schedule( new TimerTask() {
+            public void run() {
+                // log position to db if there is a new gps update
+                System.out.println("logging triggered after 10 seconds");
+
+                if (featureUpdated) {
+                    System.out.println("logging new position");
+                    table.addFeatureAsync(latestPosition);
```

Comment:
What happens if the user clicks "start logging" before `latestPosition` has been set and so it's null / `addFeatureAsync` throws? Is there a risk of the app blowing up if the buttons aren't clicked in the right order?

The UI could force the order by enabling / disabling at each stage.

Reply:
Agreed, there it lots of scope to mess up the app....   more robust UI will be added